File: shared/java/src/com/pebblefields/shared/PathAnalyzer.java
package com.pebblefields.shared;

import java.util.ArrayList;
import java.util.List;

import org.pebblefields.pebbleobject.Pebble;
import com.pebblefields.shared.interfaces.IServerData;



/**
 * PathAnalyzer gets infomation about the path 
 * @author Ryan Tsunoda
 *
 */
public class PathAnalyzer {
	private String path;
	private List<ItemType> itemTypes;				//the ItemType(typePath) so that getField, getAppIndex can be used
	private List<Pebble> fieldTypes;    			//the types as in their parent definition (field.type)
	private List<String> docPaths;					//list of doc paths (non doc members are skipped), only useful for the last one
	
	
	
	public PathAnalyzer(String appPath) {
		this.path = appPath;
		this.itemTypes = new ArrayList<ItemType>();    
		this.fieldTypes = new ArrayList<Pebble>();	
		this.docPaths = new ArrayList<String>();
		this.initComponents();
	}
	/**
	 * Set forms, docPaths
	 */
	private void initComponents() {

		String[] elements = this.path.split("\\.");
		String element = elements[0];
		Pebble top = DocModel.getDoc(element);//get top app
		this.docPaths.add(element);
		ItemType currentForm = new ItemType(TypeReferences.INDEX_ITEM);//top.getFormKey());
		
		this.itemTypes.add(currentForm);
		this.fieldTypes.add(currentForm.getCopy("."));
		
		int i = 1;
		while (i < elements.length) {
			element = elements[i];
			Pebble field = currentForm.getField(element);
			Pebble type = field.get("type");
			
			//ARRAY
			if (type.getRef(".").endsWith("standard.types.array")) { //.equals(TypeReferences.ARRAY)) {
				Pebble arrayFormId = type.get("arrayFormId");
				String instanceId = "";

				for (int j = 0; j < i + 1; j++) {
					instanceId += elements[j] + ".";
				}
				instanceId = PebbleUtils.rStrip(instanceId, ".");
				this.itemTypes.add(new ItemType(type.getRef("."))); 
				this.fieldTypes.add(type);

				if (!instanceId.equals(this.path)) {
					//skip the array element (if we didnt we would put array as currentForm but the next currentForm.getField(element) wouldn't work)
					//here we are skipping to the arrayFormId and incrementing (as the collectionPath was added to the forms list)
					instanceId += "." + elements[i + 1];
					if (type.getBool("isArrayCollection")){ 
						this.docPaths.add(instanceId);
					} 
					currentForm = new ItemType(arrayFormId.getRef("."));
					this.itemTypes.add(currentForm);
					this.fieldTypes.add(currentForm.getCopy("."));
					i += 1; 
				} 
			} else if (type.getRef(".").endsWith("standard.types.itemRelRef")){ //equals(TypeReferences.REFERENCE)) {
				//REFERENCE TO TYPE
				String pathToRef = "";
				for (int j = 0; j <= i; j++) {
					pathToRef += elements[j] + ".";
				}
				pathToRef = pathToRef.substring(0, pathToRef.lastIndexOf("."));
				String relPath = this.getRelPath(pathToRef);
				//type is in the instance.rs.k
				Pebble doc = this.getLastDoc();
				currentForm = new ItemType(doc.getRef(relPath));
				
				String configgedType = type.getRef("configgedType");
				if (configgedType != null && !configgedType.isEmpty()) {
					if (configgedType.startsWith("$reference.")){
						configgedType = configgedType.replace("$reference.", "");
						String typePath = currentForm.getRef(configgedType);
						if (typePath != null) {
							currentForm = new ItemType(typePath);
						}
					} else if (configgedType.equals("$reference")) {
						
					} else {
						currentForm = new ItemType(configgedType);
					}
				} 
				this.itemTypes.add(currentForm);
				this.fieldTypes.add(currentForm.getCopy("."));
			} else {
				currentForm = new ItemType(type.getRef("."));
				this.itemTypes.add(currentForm);
				this.fieldTypes.add(currentForm.getCopy("."));
			}
			i += 1;
		}
	}
	public String getPath() {
		return this.path;
	}
	public Pebble getLastDoc() {
		return DocModel.getDoc(this.getLastDocPath());
	}
	public List<String> getDocPaths() {
		return this.docPaths;
	}
	public String getLastDocPath() {
		return this.docPaths.get(this.docPaths.size() - 1);
	}
	public String getParentDocPath() {
		return this.docPaths.get(this.docPaths.size() - 2);
	}
	public Pebble getLastFieldType() {
		return this.fieldTypes.get(this.fieldTypes.size() - 1);
	}
	public List<ItemType> getItemTypes() {
		return this.itemTypes;
	}
	public ItemType getParentItemType() {
		return this.itemTypes.get(this.itemTypes.size() - 2);
	}
	public ItemType getLastItemType() {
		return this.itemTypes.get(this.itemTypes.size() - 1);
	}
	public String getRelPath() {
		return this.getRelPath(this.getPath());
	}
	public String getRelPath(String fullPath) {
		return PebbleUtils.lStrip(fullPath.replace(this.docPaths.get(this.docPaths.size() - 1), ""), ".");
	}
	public Pebble getItem() {
		String relPath = this.getRelPath();
		Pebble doc = this.getLastDoc();
		if (!relPath.isEmpty()) {
			return doc.get(relPath);
		} else {
			return doc;
		}
	}
	/**
	 * 
	 * @param actual
	 * @param pathMatch (spec for matching aaa.bbb.*.
	 * @return
	 */
	public static boolean comparePath(String actual, String pathMatch) {
		actual = actual.trim();
		pathMatch = pathMatch.trim();
		String[] actualElements = actual.split("\\.");
		String[] pathMatchElements = pathMatch.split("\\.");
		if (actualElements.length == pathMatchElements.length) {
			for (int i = 0; i < pathMatchElements.length; i++) {
				String match = pathMatchElements[i];
				String act = actualElements[i];
				if (match.equals(act) || match.equals("*")) {
					continue;
				} else {
					return false;
				}
			}
			//if gets here all elements matched
			return true;
		} else {
			return false;
		}
		
	}
}

File: shared/python/src/shared/pathanalyzer.py
from pebbleobject import pebble
import pebbleutils
import itemtype
import docmodel

class PathAnalyzer(object):

	def __init__(self, appPath):
		self.path = appPath
		self.itemTypes = []
		self.fieldTypes = []
		self.docPaths = []
		self.initComponents()
	
	def initComponents (self) :
	
		elements = self.path.split(".")
		element = elements[0]
		top = docmodel.DocModel.getDoc(element)#get top app
		self.docPaths.append(element)
		currentForm = itemtype.ItemType(pebble.shared.TypeReferences.INDEX_ITEM)#top.getFormKey())
	
		self.itemTypes.append(currentForm)
		self.fieldTypes.append(currentForm.getCopy("."))
	
		i = 1
		while i < len(elements):
			element = elements[i]
			field = currentForm.getField(element)
			type = field.get("type")
	
			#ARRAY
			if type.getRef(".").endswith("standard.types.array") : #.equals(TypeReferences.ARRAY)) {
				arrayFormId = type.get("arrayFormId")
				instanceId = ""
	
				for j in range(i + 1):
					instanceId += elements[j] + "."
				
				instanceId = instanceId.rstrip(".")
				self.itemTypes.append(itemtype.ItemType(type.getRef("."))) 
				self.fieldTypes.append(type)
	
				if instanceId != self.path :
					#skip the array element (if we didnt we would put array as currentForm but the next currentForm.getField(element) wouldn't work)
					#here we are skipping to the arrayFormId and incrementing (as the collectionPath was added to the forms list)
					instanceId += "." + elements[i + 1]
					if type.getBool("isArrayCollection"): 
						self.docPaths.append(instanceId)
					 
					currentForm = itemtype.ItemType(arrayFormId.getRef("."))
					self.itemTypes.append(currentForm)
					self.fieldTypes.append(currentForm.getCopy("."))
					i += 1 
				 
			elif type.getRef(".").endswith("standard.types.itemRelRef"): #equals(TypeReferences.REFERENCE)) {
				#REFERENCE TO TYPE
				pathToRef = ""
				for j in range(i + 1):
					pathToRef += elements[j] + "."
				
				pathToRef = pathToRef.substring(0, pathToRef.lastIndexOf("."))
				relPath = self.getRelPath(pathToRef)
				#type is in the instance.rs.k
				doc = self.getLastDoc()
				currentForm = itemtype.ItemType(doc.getRef(relPath))
	
				configgedType = type.getRef("configgedType")
				if configgedType is not None and configgedType != "" :
					if configgedType.indexOf("$reference.") == 0:
						configgedType = configgedType.replace("$reference.", "")
						typePath = currentForm.getRef(configgedType)
						if typePath is not None :
							currentForm = itemtype.ItemType(typePath)
						
					elif configgedType == "$reference" :
						pass	
					else:
						currentForm = itemtype.ItemType(configgedType)
				 
				self.itemTypes.append(currentForm)
				self.fieldTypes.append(currentForm.getCopy("."))
			else:
				currentForm = itemtype.ItemType(type.getRef("."))
				self.itemTypes.append(currentForm)
				self.fieldTypes.append(currentForm.getCopy("."))
			
			i += 1
		
	
	def getPath (self) :
		return self.path
	
	def getLastDoc (self) :
		return docmodel.DocModel.getDoc(self.getLastDocPath())
	
	def getDocPaths (self) :
		return self.docPaths
	
	def getLastDocPath (self) :
		return self.docPaths[len(self.docPaths) - 1]
	
	def getParentDocPath (self) :
		return self.docPaths[len(self.docPaths) - 2]
	
	def getLastFieldType (self) :
		return self.fieldTypes[len(self.fieldTypes) - 1]
	
	def getItemTypes (self) :
		return self.itemTypes
	
	def getParentItemType (self) :
		return self.itemTypes[len(self.itemTypes) - 2]
	
	def getLastItemType (self) :
		return self.itemTypes[len(self.itemTypes) - 1]
	
	def getRelPath (self, fullPath) :
		if fullPath is None :
			fullPath = self.getPath()
		
		return fullPath.replace(self.docPaths[len(self.docPaths) - 1], "").lstrip(".")
	
	def getItem (self) :
		relPath = self.getRelPath()
		doc = self.getLastDoc()
		if relPath != "" :
			return doc.get(relPath)
		else:
			return doc
		
	
